JavaPro Задание #8 (Spring Boot и PostgreSQL)

klaus1776 opened the pull request:
Александр, добрый день!

Выполнил итоговое Задание #8 курса JavaPro.

Реализовал REST микросервис лимитов.
В результате приложение состоит из трех сервисов:
1. Сервиса продуктов - ProductCore.
2. Сервиса платежей - PaymentCore.
3. Сервиса лимитов - LimitsCore.

Поскольку в задании не было написано какой именно сервис должен реализовывать платежи с контролем по лимитам реализовал так:
Я решил, что по логике, лимиты это просто фича для платежей, поэтому реализуются платежи сервисом платежей PaymentCore.
В результате у меня в сервисе платежей реализовано два Bean-а RestTemplate, один подключается к сервису продуктов, второй - к сервису лимитов.
Настроек с application.yml соответственно тоже две.
Поиск продукта с которого списываются средства по платежу осуществляется по номеру счета, а не по id - так мне показалось логичнее.
Номера счетов уникальные (constraint на уровне БД).
Контроль открытия (если нет) и уменьшения лимита происходит в классе PaymentsService сервиса платежей PaymentCore, в методе PaymentsService.makePayment.
При этом уменьшение лимита происходит на сумму успешно реализованного платежа после завершения платежа, таким образом откат не требуется.
Если платеж по какой-либо причине не прошел (вывалилась ошибка), то уменьшения лимита просто не происходит из-за ошибки проведения платежа.
Сервис лимитов LimitsCore., таким образом просто предоставляет функционал по лимитам.

Проверьте, пожалуйста.

klaus1776 committed:
JavaPro Задание #7 (Spring Boot и PostgreSQL):
Сделал два сервиса:
Сервис платежей "PaymentCore" и сервис продуктов "ProductCore".
Добавил в оба сервиса обработчики ошибок.
Всю бизнес-логику проведения платежей, т.е. проверку существования продукта и достаточность средств на нем, разместил в PaymentCore.
Для выбора продукта сделал на стороне PaymentCore загрузку всех продуктов клиента по id клиента.
На стороне PaymentCore так же сделал возможность загрузки информации о продукте по id продукта и по номеру счета продукта (account).

klaus1776 committed:
JavaPro Задание #8 (Spring Boot и PostgreSQL):

Выполнил итоговое Задание #8 курса JavaPro.

Реализовал REST микросервис лимитов.
В результате приложение состоит из трех сервисов:
1. Сервиса продуктов - ProductCore.
2. Сервиса платежей - PaymentCore.
3. Сервиса лимитов - LimitsCore.

Поскольку в задании не было написано какой именно сервис должен реализовывать платежи с контролем по лимитам реализовал так:
Я решил, что по логике, лимиты это просто фича для платежей, поэтому реализуются платежи сервисом платежей PaymentCore.
В результате у меня в сервисе платежей реализовано два Bean-а RestTemplate, один подключаетс к сервису продуктов, второй - к сервису лимитов.
Настроек с application.yml соответственно тоже двею
Поиск продукта с которого списываются средства по платежу осуществляется по номеру счета, а не по id - так мне показалось логичнее.
Номера счетов уникальные (constraint на уровне БД).
Контроль открытия (если нет) и уменьшения лимита происходит в класее PaymentsService сервиса платедей PaymentCore, в методе PaymentsService.makePayment.
При этом уменьшение лимита происходит на сумму успешно реализованного платежа после завершения платежа, таким образом откат не требуется.
Если платеж по какой-либо причине не прошел (вывалилась ошибка), то уменьшения лимита просто не происходит из-за ошибки проведения платежа.
Сервис лимитов LimitsCore., таким образом просто предоставляет функционал по лимитам.

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/dto/UsersLimitDto.java:
```diff
@@ -0,0 +1,6 @@
+
+import edu.innotech.entity.UsersLimit;
+
+public record UsersLimitDto(UsersLimit usersLimit) {
```
нет! entity не используется в ответах! для этого есть мапперы

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/entity/UsersLimit.java:
```diff
@@ -0,0 +1,33 @@
+@Getter
+@Setter
+@Entity
+@Table(name = "userslimits")
```
таблицу может тоже назвать в snake_case? hiber по умолчанию работает так, поля у вас названы так же… на подумать на будущее

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/service/LimitsService.java:
```diff
@@ -0,0 +1,87 @@
+    private final String limitsStartDate;
+    private final Double limitsValue;
+
+    // Метод, обновляющий лимиты пользователей в 00:00:00
```
нет, тут нужен был [@sheduler](https://www.baeldung.com/spring-scheduled-tasks)

Alexandr056 replied:
нет, здесть нужен был [sheduler](https://www.baeldung.com/spring-scheduled-tasks)

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/service/LimitsService.java:
```diff
@@ -0,0 +1,87 @@
+        );
+    }
+
+    public LimitsService(LimitsRepository limitsRepository,
```
конструктор в начало

klaus1776 replied:
Перенес

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/service/LimitsService.java:
```diff
@@ -0,0 +1,87 @@
+
+    public LimitsService(LimitsRepository limitsRepository,
+                         //RestTemplate restTemplate,
+                         @Value("${service.limits-client-products}") String limitsProducts,
```
лимит не по продуктам, название параметра путает

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/service/LimitsService.java:
```diff
@@ -0,0 +1,87 @@
+    public LimitsService(LimitsRepository limitsRepository,
+                         //RestTemplate restTemplate,
+                         @Value("${service.limits-client-products}") String limitsProducts,
+                         @Value("${service.limits-start-date-format}") String limitsDateFormat,
```
не нужно

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/service/LimitsService.java:
```diff
@@ -0,0 +1,87 @@
+                         //RestTemplate restTemplate,
+                         @Value("${service.limits-client-products}") String limitsProducts,
+                         @Value("${service.limits-start-date-format}") String limitsDateFormat,
+                         @Value("${service.limits-start-date}") String limitsStartDate,
```
не нужено

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/service/LimitsService.java:
```diff
@@ -0,0 +1,87 @@
+                         @Value("${service.limits-client-products}") String limitsProducts,
+                         @Value("${service.limits-start-date-format}") String limitsDateFormat,
+                         @Value("${service.limits-start-date}") String limitsStartDate,
+                         @Value("${service.limits-value}") Double limitsValue) {
```
а вот и еще лимит, а должен быть один

klaus1776 replied:
Убрал

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/service/LimitsService.java:
```diff
@@ -0,0 +1,87 @@
+        return usersLimitDto;
+    }
+
+    public void updateLimitValue(Long userId, Double limit) {
```
нет, не просто обновление, нужен метод на уменьшение лимита и восстановление:, который будет вызываться из контроллера

klaus1776 replied:
Добавил два метода работы с лимитами: 1) метод уменьшения лимита decreaseLimitValue и 2) метод восстановления (увеличения) лимита recoveryLimitValue.

Alexandr056 reviewed (changes requested):
проверяю только limitsCore остальное не нужно! один сервис, его надо немного доработать

klaus1776 committed:
JavaPro Задание #8 (Spring Boot и PostgreSQL):

Исправил замечания Александра:
1. Добавил в пакет dto маппер UserLimitMapper (наверно надо было для него отдельный пакет создать) и преобразовал DTO-шку UsersLimitDto из record в class. Добавил мапинг в сервис LimitsService в метод findLimitByUserId.
2. Заменил timer.schedule при обновлении лимитов на на класс DailyTaskScheduler с методом executeUpdateUsersLimitsTask аннотированным @Scheduled(cron = "0 0 0 * * ?").
3. Перенес конструктор класса LimitsService в начало кода класса.
4. Убрал лишние настройки из кода и из application.yml.
5. Добавил два метода работы с лимитами: 1) метод уменьшения лимита decreaseLimitValue и 2) метод восстановления (увеличения) лимита recoveryLimitValue.

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/controller/LimitsController.java:
```diff
@@ -23,4 +23,16 @@ public UsersLimitDto checkUserLimit(@RequestParam("userId") Long userId) {
     }
+
+    // Метод, уменьшающий лимит клиента
+    @PostMapping(value = "/limitdec")
```
limit/dec

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/dto/UserLimitMapper.java:
```diff
@@ -0,0 +1,16 @@
+
+import org.modelmapper.ModelMapper;
+
+public class UserLimitMapper {
```
можно сделать бином

Alexandr056 commented on LimitsCore/src/main/java/edu/innotech/taskScheduler/DailyTaskScheduler.java:
```diff
@@ -0,0 +1,20 @@
+        this.limitsService = limitsService;
+    }
+
+    // Метод, обновляющий лимиты пользователей в 00:00:00
```
да!

Alexandr056 reviewed (approved):
хрошая работа над ошибками
